new tool icescreen

thomas-lacroix opened the pull request:
ICEscreen is a bioinformatic pipeline for the detection and annotation of ICEs (Integrative and Conjugative Elements) and IMEs (Integrative and Mobilizable Elements) in Firmicutes genomes.

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

bgruening commented on tools/icescreen/icescreen.xml:
```diff
@@ -0,0 +1,100 @@
+        #import os
+        #set $gbname = os.path.splitext($genome.element_identifier)[0]
+        #set $tmpfolder = mkdtemp()
+        mkdir '${tmpfolder}'/source_genbank &&
```
you can just create a folder in the current working dir ... `mkdir ./mytempfolder/` and avoid the mkdtemp

thomas-lacroix replied:
Done

bgruening commented on tools/icescreen/icescreen.xml:
```diff
@@ -0,0 +1,100 @@
+    <command detect_errors="exit_code"><![CDATA[
+        #from tempfile import mkdtemp
+        #import os
+        #set $gbname = os.path.splitext($genome.element_identifier)[0]
```
can we avoid element_identifier and just use a static name? `galaxy_genbank_` or something? Is there an advantage to use the history name?

thomas-lacroix replied:
I don't see how to avoid using the element_identifier and the $gbname, the $gbname is part of the path of the output files generated by the tool. The tool was designed to run simultaneously on multiple GenBank files, this is why we included the $gbname in the path of the output files originally. But maybe I am missing your point ?

bgruening commented on tools/icescreen/icescreen.xml:
```diff
@@ -0,0 +1,100 @@
+        </section>
+    </inputs>
+    <outputs>
+        <data format="txt" name="log" label="${tool.name} on ${on_string}: log file" />
```
are all outputs needed? Maybe you can make them optional and let the user choose a few? zip for example looks like one output that is not always needed.

thomas-lacroix replied:
5 output files are now optional, I also updated the test section accordingly.

bgruening reviewed (commented):
Cool, thanks @thomas-lacroix ... I added a few questions inline.

thomas-lacroix commented:
Thanks for your review @bgruening , good points. I will make the changes and push them.

thomas-lacroix commented:
Almost... 1 test failed: "Combine chunked test results (3.7)". This error is a bit cryptic to me ;) Any idea what it means and how to fix it @bgruening ? All my tests were successful locally (planemo lint, planemo test, planemo serve, planemo shed_lint).

thomas-lacroix commented:
I dug a little deeper into this thanks to discussions on gitter. I downloaded the 2 artifact and looked at "All tools test results" and "Tool test output 0". It seems that the 7 tests all failed: `Job in error state.. tool_id: icescreen, exit_code: 1, stderr: .`
I was confused that "Test tools (O, 3.7)" was successfull and thought it meant the tests were OK, but when looking at the log of "Planemo test" :
`2021-12-10 08:33:47,293 INFO  [galaxy.tool_util.output_checker] Job error detected, failing job. Reasons are [{'type': 'exit_code', 'desc': 'Fatal error: Exit code 1 ()', 'exit_code': 1, 'code_desc': '', 'error_level': 3}]
FAIL`
But this error is not very explicit and I still don't know how to move forward to resolve it. I can not reproduce the failure locally as all the test pass when I do :
`planemo --verbose test --install_galaxy /PATH/TO/icescreen.xml`
I will ask on gitter monday.

bgruening commented:
@thomas-lacroix I added one commit with a few cleanups and my suggestion about the element_identifier. I can revert that if this does not work for you.

I see this error now: 
> local variable 'contents' referenced before assignment

Which looks to me like a Python error in your tool, can this be? it only happens in test 5.

![image](https://user-images.githubusercontent.com/469983/145676812-9ae209bd-9d26-4e5e-ac47-871371e378e4.png)

bgruening commented on tools/icescreen/icescreen.xml:
```diff
@@ -0,0 +1,194 @@
+                <option value="streptomyces">Streptomyces</option>
+            </param>
+        </section>
+        <section name="additional_output" title="Additional output files" expanded="False">
```
@thomas-lacroix you could do this with one small select field with multiple=true and optional=true

thomas-lacroix commented:
Thank you @bgruening for all your changes, it looks better now. I now understand your point about `$gbname = 'galaxy_genbank'`. Good idea for the selection of additional output files using multiple choice checkbox, I updated the wrapper. I also have changed the ordering of the conditional commands: dealing with the command to produce the log file needed to be before the command to generate the zip file as the latter features a `cd`. I added a test in this regard.

thomas-lacroix commented:
With regard to the problem with test#5, I don't understand the error `local variable 'contents' referenced before assignment` as there is no variable named 'contents' in our scripts and `grep -R 'contents' .` return empty. The test #5 deals with the output_zip_all_files parameter, this parameter creates an extra zip file of the results folder but does not affect how the tool itself runs. When I test icescreen.xml in a local installation of Galaxy (http://localhost:8080/), the zip output looks fine when the option is required. I was somewhat able to reproduce the failure locally, but with the error `
'NoneType' object has no attribute 'read'
Traceback (most recent call last):
  File "/tmp/tmpfu4cxf42/galaxy-dev/test/functional/test_toolbox.py", line 98, in test_tool
    self.do_it(tool_version=tool_version, test_index=test_index)
  File "/tmp/tmpfu4cxf42/galaxy-dev/test/functional/test_toolbox.py", line 35, in do_it
    verify_tool(tool_id, self.galaxy_interactor, resource_parameters=resource_parameters, test_index=test_index, tool_version=tool_version, register_job_data=register_job_data)
  File "/tmp/tmpfu4cxf42/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1131, in verify_tool
    raise e
  File "/tmp/tmpfu4cxf42/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1127, in verify_tool
    job_stdio = _verify_outputs(testdef, test_history, jobs, data_list, data_collection_list, galaxy_interactor, quiet=quiet)
  File "/tmp/tmpfu4cxf42/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1301, in _verify_outputs
    raise JobOutputsError(found_exceptions, job_stdio)
galaxy.tool_util.verify.interactor.JobOutputsError: 'NoneType' object has no attribute 'read'
Exit Code: 0
`
The error was due to `<has_archive_member path=".*/NC_004668_137848_164286_detected_SP_withMEIds.tsv">` but it works when I change it to `<has_archive_member path=".*/*_detected_SP_withMEIds.tsv">`.

thomas-lacroix commented:
The pipeline of tests still fails at "Combine chunked test results (3.7)", and looking in details at "Planemo test" it seems all the test fail:
```
 Traceback (most recent call last):
   File "/tmp/tmp6z4yumfi/galaxy-dev/test/functional/test_toolbox.py", line 98, in test_tool
     self.do_it(tool_version=tool_version, test_index=test_index)
   File "/tmp/tmp6z4yumfi/galaxy-dev/test/functional/test_toolbox.py", line 35, in do_it
     verify_tool(tool_id, self.galaxy_interactor, resource_parameters=resource_parameters, test_index=test_index, tool_version=tool_version, register_job_data=register_job_data)
   File "/tmp/tmp6z4yumfi/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1131, in verify_tool
     raise e
   File "/tmp/tmp6z4yumfi/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1127, in verify_tool
     job_stdio = _verify_outputs(testdef, test_history, jobs, data_list, data_collection_list, galaxy_interactor, quiet=quiet)
   File "/tmp/tmp6z4yumfi/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1301, in _verify_outputs
     raise JobOutputsError(found_exceptions, job_stdio)
 galaxy.tool_util.verify.interactor.JobOutputsError: Job in error state.. tool_id: icescreen, exit_code: 1, stderr: .
 Job in error state.. tool_id: icescreen, exit_code: 1, stderr: .
```
But why, I dont have those errors locally ?

bgruening commented:
Are you running it locally with `planemo test --docker`?

thomas-lacroix commented:
You are right, when I use `planemo test --docker` locally, all the tests fail in a similar fashion as the pipeline remotely:
```
  File "/tmp/tmp0scxw5s9/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1301, in _verify_outputs
    raise JobOutputsError(found_exceptions, job_stdio)
galaxy.tool_util.verify.interactor.JobOutputsError: Job in error state.. tool_id: icescreen, exit_code: 127, stderr: /bin/bash: /tmp/tmp0scxw5s9/job_working_directory/000/16/tool_script.sh: No such file or directory
```
I didn't find any usefull hint on how to resolve this problem on the internet. Do you know how to move forward past that error and see the "real" results of the `planemo test` ?

bgruening commented:
I assume a dependency is missing that is installed on your machine, but not in the containers. You can look at your local containers and try this one out manually.

thomas-lacroix commented:
I tried to reproduce the error by running docker locally as follow:
```
docker run -dit --name icescreen_test -d quay.io/biocontainers/icescreen:1.0.0--py310hdfd78af_0
docker attach icescreen_test
cd
python3 -m venv planemo
. planemo/bin/activate
pip install planemo
pip install -U planemo
docker cp /PATH/TO/icescreen.xml CONTAINER_ID:/root
docker cp /PATH/TO/.shed.yml CONTAINER_ID:/root
docker cp /PATH/TO/test-data CONTAINER_ID:/root
cd /root
planemo --verbose test --install_galaxy icescreen.xml
```
However after launching the last command nothing is printed on the terminal and my free disk space slowly goes from 56GB to 0MB. In other work, this process is using all my disk space in an unusual fashion.

thomas-lacroix commented:
However it is unlikely that the errors come from a missing dependency related to the tool as:
 - setting up a VM with minimal ubuntu 20.04, git, and miniconda is sufficient to guarantee successful installation of the tool via conda (`conda create -y --name icescreen_env icescreen -c conda-forge -c bioconda`) or from source.
 - installing planemo and running `planemo test --install_galaxy` on the above VM results in all 8 tests passing (no failure)

@bgruening , is it be possible that those errors are related to planemo --docker ?

bgruening commented:
Unlikely, its working for us here and on production servers for years. I will try to have a look.

bgruening commented:
I see that here.

```
h3m index is missing, remove DONE flag file and recompute indexes
failed to open SSI index /usr/local/bin/detection_SP/database/hmmdb/SP_profiles/Relaxase_profile_MOBT.hmm.h3i
h3i index is missing, remove DONE flag file and recompute indexes
failed to open SSI index /usr/local/bin/detection_SP/database/hmmdb/SP_profiles/Relaxase_profile_MOBT.hmm.h3i
h3f index is missing, remove DONE flag file and recompute indexes
failed to open SSI index /usr/local/bin/detection_SP/database/hmmdb/SP_profiles/Relaxase_profile_MOBT.hmm.h3i
h3p index is missing, remove DONE flag file and recompute indexes
failed to open SSI index /usr/local/bin/detection_SP/database/hmmdb/SP_profiles/Relaxase_profile_MOBT.hmm.h3i
All indexes files exist but not flag file
touch: /usr/local/bin/detection_SP/database/Relaxase_profile_MOBT_indexes.DONE: Permission denied
[Wed Dec 15 00:26:30 2021]
Error in rule hmmpress_SP:
    jobid: 11
    output: /usr/local/bin/detection_SP/database/Relaxase_profile_MOBT_indexes.DONE
    shell:
        
        for ext in h3m h3i h3f h3p
        do
            if [ ! -f /usr/local/bin/detection_SP/database/hmmdb/SP_profiles/Relaxase_profile_MOBT.hmm."$ext" ]
            then
                echo "$ext index is missing, remove DONE flag file and recompute indexes"
                rm -f /usr/local/bin/detection_SP/database/Relaxase_profile_MOBT_indexes.DONE
                hmmpress -f /usr/local/bin/detection_SP/database/hmmdb/SP_profiles/Relaxase_profile_MOBT.hmm && touch /usr/local/bin/detection_SP/database/Relaxase_profile_MOBT_indexes.DONE
            fi
        done
        # indexes files exist but DONE flag file does not exist
        echo "All indexes files exist but not flag file"
        touch /usr/local/bin/detection_SP/database/Relaxase_profile_MOBT_indexes.DONE
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```

And I think its this here: 
> touch: /usr/local/bin/detection_SP/database/Relaxase_profile_MOBT_indexes.DONE: Permission denied

You are writing to a folder that is not writable.

thomas-lacroix commented:
Thanks for the finding, that must be it :) I will look into it. Is there a guide I can follow to reproduce this error locally ? I couldn't come across this error with my different tries, the errors for the failure that I got were not explicit which makes the debugging process confusing. I am thinking that if I can reproduce the steps locally I can then test the different strategies to fix this bug more efficiently.

thomas-lacroix commented:
Hello !
The bug related to `/usr/local/bin/ Permission denied` has been fixed in ICEscreen release 1.0.1, the generation of makeblastdb and HMMPress files are now performed during build time instead of run time. My local tests for the fixing of this bug are conclusive and I pushed the new version to bioconda.
My local tests `planemo lint`, `planemo serve`, and `planemo test --install_galaxy` are conclusive. However all the tests are failing again with my local `planemo test --docker --install_galaxy`... The log file doesn't provide clear hints to what went wrong (even with option `planemo --verbose`):
`Job in error state.. tool_id: icescreen, exit_code: 127, stderr: /bin/bash: /tmp/tmpq3n2ezhb/job_working_directory/000/2/tool_script.sh: No such file or directory`
I would like to run the `planemo test --docker --install_galaxy` on the remote pipeline to see if this error goes away and is due to my local configuation of Docker.

bgruening commented:
I triggered the test here, thanks @thomas-lacroix

thomas-lacroix commented:
This looks better, only the tests #5 et #8 failed, and both are related to testing the zip output option and the error message is `zip: command not found`. I added a `<requirement type="package" version="3.0">zip</requirement>`. I can not use my local `planemo test --docker` to confirm the fix as all my tests fail locally, I don't know why. Lets see if this works out.

thomas-lacroix commented:
It is unexpected, all tests failed because of the addition of the zip requirement whereas only 2 were failing prior to that. I didn't find any useful error message in the logs, the package zip seems to install just fine. that's a puzzle...

bgruening commented on tools/icescreen/icescreen.xml:
```diff
@@ -1,11 +1,12 @@
-<tool id="icescreen" name="ICEscreen" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="20.09">
```
you have removed the profile as well

thomas-lacroix commented:
Thanks @bgruening , I removed the profile unintentionally but adding it back worked, all tests are passing now ! Is there anything else to do before merging ?

bgruening commented:
> Thanks @bgruening , I removed the profile unintentionally but adding it back worked, all tests are passing now ! Is there anything else to do before merging ?

Yes, this is because we are now isolating the HOME more agressivley. You seem to someone rely on this directory. Glad it works now. Let me review this PR, one final time.

bgruening reviewed (approved):
Looks good to me. I think you do not need the sections as you only have one parameter included in those, but that is up to you :)

Thanks for your contribution. Anything you want to add?

thomas-lacroix commented:
Great ! We can leave the section for now, that way the advanced parameter is hidden by default. I don't have anything to add, I think we are good to go for the merge :) Thanks again for your tips and your help with the debugging !

bgruening commented:
Cool, thanks again @thomas-lacroix! This tool will now be automatically uploaded to the toolshed.
